Wikipedia/Code/ArticleViewController+Notifications.swift

```diff
@@ -0,0 +1,182 @@
+    }
+    
+    /// Adds the lead image view to the web view's scroll view and configures the associated constraints
+    func setupLeadImageView() {
+        webView.scrollView.addSubview(leadImageContainerView)
+
+        let leadingConstraint =  leadImageContainerView.leadingAnchor.constraint(equalTo: webView.leadingAnchor)
+        let trailingConstraint =  webView.trailingAnchor.constraint(equalTo: leadImageContainerView.trailingAnchor)
+        let topConstraint = webView.scrollView.topAnchor.constraint(equalTo: leadImageContainerView.topAnchor)
+        let imageTopConstraint = leadImageView.topAnchor.constraint(equalTo:  leadImageContainerView.topAnchor)
+        imageTopConstraint.priority = UILayoutPriority(rawValue: 999)
+        let imageBottomConstraint = leadImageContainerView.bottomAnchor.constraint(equalTo: leadImageView.bottomAnchor, constant: leadImageBorderHeight)
+        NSLayoutConstraint.activate([topConstraint, leadingConstraint, trailingConstraint, leadImageHeightConstraint, imageTopConstraint, imageBottomConstraint, leadImageLeadingMarginConstraint, leadImageTrailingMarginConstraint])
+        
+        articleAsLivingDocController.setupLeadImageView()
+    }
+    
+    func setupPageContentServiceJavaScriptInterface(with completion: @escaping () -> Void) {
+        guard let siteURL = articleURL.wmf_site else {
+            DDLogError("Missing site for \(articleURL)")
+            showGenericError()
+            return
+        }
+        
+        // Need user groups to let the Page Content Service know if the page is editable for this user
+        authManager.getLoggedInUser(for: siteURL) { (result) in
+            assert(Thread.isMainThread)
+            switch result {
+            case .success(let user):
+                self.setupPageContentServiceJavaScriptInterface(with: user?.groups ?? [])
+            case .failure:
+                DDLogError("Error getting userinfo for \(siteURL)")
+                self.setupPageContentServiceJavaScriptInterface(with: [])
+            }
+            completion()
+        }
+    }
+    
+    func setupPageContentServiceJavaScriptInterface(with userGroups: [String]) {
+        let areTablesInitiallyExpanded = altTextExperimentViewModel != nil ? true : UserDefaults.standard.wmf_isAutomaticTableOpeningEnabled
+
+        messagingController.shouldAttemptToShowArticleAsLivingDoc = articleAsLivingDocController.shouldAttemptToShowArticleAsLivingDoc
+
+        messagingController.setup(with: webView, languageCode: articleLanguageCode, theme: theme, layoutMargins: articleMargins, leadImageHeight: leadImageHeight, areTablesInitiallyExpanded: areTablesInitiallyExpanded, userGroups: userGroups)
+    }
+    
+    func setupToolbar() {
+        enableToolbar()
+        toolbarController.apply(theme: theme)
+        toolbarController.setSavedState(isSaved: article.isAnyVariantSaved)
+        setToolbarHidden(false, animated: false)
+    }
```

Comment:
There are several `setup`-y method names, here and elsewhere. Maybe they could live in their own +Setup extensions file?